riscv-page-tables/src/page_table.rs

```diff
@@ -737,6 +792,33 @@ impl<T: PagingMode> PageTableInner<T> {
             _ => Err(Error::PageNotInvalidated),
         }
     }
+
+    /// Promotes the table entry at `vaddr` as a valid leaf.
+    ///
+    /// # Safety
+    ///
+    /// The caller must guarantee that `paddr` references a page uniquely owned by the root
+    /// `GuestStagePageTable`.
+    unsafe fn promote_pte(
+        &mut self,
+        vaddr: RawAddr<T::MappedAddressSpace>,
+        paddr: SupervisorPageAddr,
+        page_size: PageSize,
+    ) -> Result<(LeafPte<T>, Page<InternalClean>)> {
+        let mut entry = PageTable::from_root(self).entry_for_addr_mut(vaddr);
+        use TableEntryType::*;
+        while let Table(t) = entry {
+            if t.level().leaf_page_size() != page_size {
+                entry = t.table().entry_for_addr_mut(vaddr);
+            } else {
```

Comment:
nit: this is a bit easier to read if you do `if <page size matches> { promote } else { next PTE }` than the other way around.